### PR TITLE
Feature/pytest-and-mocking

### PR DESCRIPTION
**Pull request by vesnikos**
in this PR:
 - added mock functionality to mimic s3 functionality through the moto library
 - added airbnb_db mock fixture
 - tests on how it's used
 - updated airflow to 2.9.1 
 - added aws and postgres airflow packages
 - added test option in the makefile

### CODE REVIEW

**Comment by ErmiaUOG** [comment]
Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.

We always need to have git hygen. So we always need to do git rebase to remove all wip commits.

File: tests/conftest.py [hunk 1]
```diff
@@ -1,4 +1,49 @@
...
+
 import pytest
+from moto.moto_server.threaded_moto_server import ThreadedMotoServer
+from s3fs.core import S3FileSystem
+
+# from s3fs import S3FileSystem
+
+port = 5555
+entpoint_s3 = "http://127.0.0.1:%s" % port
+bucket_name = "test-bucket"
```
Comment (ErmiaUOG) [inline tests/conftest.py:10]: Why not use env variable for them?

File: tests/conftest.py [hunk 2]
```diff
@@ -1,4 +1,49 @@
...
+
+
+@pytest.fixture(scope="module")
+def s3_base():
+    # writable local s3 system
+    os.environ["AIRFLOW_CONN_AWS_DEFAULT"] = (
+        """{"conn_type": "aws", "extra": {"endpoint_url": "http://127.0.0.1:5555"}}"""
+    )
+
+    server = ThreadedMotoServer(ip_address="127.0.0.1", port=port)
```
Comment (ErmiaUOG) [inline tests/conftest.py:25]: Create var for the ip address too?

File: pyproject.toml [hunk 3]
```diff
@@ -8,8 +8,13 @@ readme = "README.md"
 
 [tool.poetry.dependencies]
 python = ">=3.10,<3.13"
-apache-airflow = "^2.9.0"
+apache-airflow = {extras = ["amazon"], version = "^2.9.1"}
 pendulum = "^3.0.0"
+moto = "^5.0.7"
+s3fs = "^2024.5.0"
+flask-cors = "^4.0.1"
```
Comment (ErmiaUOG) [inline pyproject.toml:15]: why?

**Comment by vesnikos** [comment]
> Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.
> 
> We always need to have git hygen. So we always need to do git rebase to remove all wip commits.

agree. usual approach here would be do squash all commits in this PR into one, giving the author the change to rework the description and getting into as much detail is needed. In this case the wip commits will be removed.
no need to retain the history of the PR from the authors point of view

File: pyproject.toml [hunk 4]
```diff
@@ -8,8 +8,13 @@ readme = "README.md"
...
+apache-airflow-providers-postgres = "^5.11.0"
```
Comment (ErmiaUOG) [inline pyproject.toml:16]: Why?

File: tests/conftest.py (hunk 1 above)
Comment (vesnikos) [reply tests/conftest.py:10]: it's for testing; the `entpoint_s3` variable is used bellow to set up the testing environment

File: pyproject.toml (hunk 3 above)
Comment (vesnikos) [reply pyproject.toml:15]: moto is aws mock library; flask-cors is needed either from s3fs or moto (don't remember)

**Comment by ErmiaUOG** [comment]
> > Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.
> > We always need to have git hygen. So we always need to do git rebase to remove all wip commits.
> 
> agree. usual approach here would be do squash all commits in this PR into one, giving the author the change to rework the description and getting into as much detail is needed. In this case the wip commits will be removed. no need to retain the history of the PR from the authors point of view

That is also big no from me. Why squash? we want to keep the git history to know why Ermia adds dag.py. Why did he update that file?

**Comment by vesnikos** [comment]
> > > Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.
> > > We always need to have git hygen. So we always need to do git rebase to remove all wip commits.
> > 
> > 
> > agree. usual approach here would be do squash all commits in this PR into one, giving the author the change to rework the description and getting into as much detail is needed. In this case the wip commits will be removed. no need to retain the history of the PR from the authors point of view
> 
> That is also big no from me. Why squash? we want to keep the git history to know why Ermia adds dag.py. Why did he update that file?

huh? I think you misunderstood; work that the author has done at his side for a PR is being squashed. Why would you like to keep track on the intermediate steps of how the author has reached to the point that he's ready to present his work in his PR? It makes no sense to me. These are the steps that being squashed into a single commit with the appropriate message that is minted into main. 

What is happening in the main branch is not getting squashed nor will ever be without a good reason.

**Comment by vesnikos** [comment]
comments from @ErmiaUOG are reflected with https://github.com/urbanbigdatacentre/ubdc-dags/pull/30/commits/bcde224f8b03c98821535db99c534e1e294fa28d. give them a look

**Review by ErmiaUOG — APPROVED** [review]
LGTM!

**Comment by ErmiaUOG** [comment]
> > > > Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.
> > > > We always need to have git hygen. So we always need to do git rebase to remove all wip commits.
> > > 
> > > 
> > > agree. usual approach here would be do squash all commits in this PR into one, giving the author the change to rework the description and getting into as much detail is needed. In this case the wip commits will be removed. no need to retain the history of the PR from the authors point of view
> > 
> > 
> > That is also big no from me. Why squash? we want to keep the git history to know why Ermia adds dag.py. Why did he update that file?
> 
> huh? I think you misunderstood; work that the author has done at his side for a PR is being squashed. Why would you like to keep track on the intermediate steps of how the author has reached to the point that he's ready to present his work in his PR? It makes no sense to me. These are the steps that being squashed into a single commit with the appropriate message that is minted into main.
> 
> What is happening in the main branch is not getting squashed nor will ever be without a good reason.

If we have a bug next year and I get back in the file and see the general commit, I do not know why we added this line of code.
We use commit message to know why we go with this solution

**Comment by vesnikos** [comment]
> > > > > Personal opinion: In my previous company, we always used a git commit-like story. Each commitment will help us understand why we have decided on this.
> > > > > We always need to have git hygen. So we always need to do git rebase to remove all wip commits.
> > > > 
> > > > 
> > > > agree. usual approach here would be do squash all commits in this PR into one, giving the author the change to rework the description and getting into as much detail is needed. In this case the wip commits will be removed. no need to retain the history of the PR from the authors point of view
> > > 
> > > 
> > > That is also big no from me. Why squash? we want to keep the git history to know why Ermia adds dag.py. Why did he update that file?
> > 
> > 
> > huh? I think you misunderstood; work that the author has done at his side for a PR is being squashed. Why would you like to keep track on the intermediate steps of how the author has reached to the point that he's ready to present his work in his PR? It makes no sense to me. These are the steps that being squashed into a single commit with the appropriate message that is minted into main.
> > What is happening in the main branch is not getting squashed nor will ever be without a good reason.
> 
> If we have a bug next year and I get back in the file and see the general commit, I do not know why we added this line of code. We use commit message to know why we go with this solution

- if you have a bug and you found out about it in a year of then, there something else wrong in the approach in general. 
- git commit messages are **not** a place for a discussion or code commenting but rather briefly mentioned what was changed. If the reasons of the commit or PR was triggered because of a significant reason; there **should** be an issue closed or otherwise discussing it freely there.
- if there's a particularly complicated code block one should use code comments to describe what it's done.

**Pull request merged**